boruta/script.whizzml

```diff
@@ -0,0 +1,391 @@
+;;   to-names-fn: (function) Function to transform IDs to field names
+;;
+;; Output: (list) List of field names
+(define (select-by-importance f-map score to-names-fn)
+  (let (selected (filter-map (lambda (x y) (= y score)) f-map))
+    (map to-names-fn (keys selected))))
+
+;; group-importance
+;;
+;; Groups the fields according to their importance
+;;
+;; Inputs:
+;;   f-rank: (map) Map from field ID to importance flag
+;;   fields: (map) Fields information structure
+;;   f-names: (list) Fields names
+;;   objective: (string) objective field name
+;;
+;; Output:  (map) Map from categories (important, unimportant or neutral) to
+;;                list of field names
+;;
+(define (group-importance f-rank fields f-names objective)
+  (let (to-names (lambda (x) (get-in fields [x "name"]))
+        important (map to-names
+                       (keys (filter-map (lambda(x y) (> y 0))
+                                         f-rank)))
+        important (append important objective)
+        neutral (map to-names
+                     (keys (filter-map (lambda(x y) (= y 0))
+                                       f-rank)))
+        unimportant (filter (lambda (x) (not (or (member? x important)
+                                                 (member? x neutral))))
+                            f-names))
+    (assoc (assoc {}
+                  "important" important)
+                  "unimportant" unimportant
+                  "neutral" neutral)))
+
+;; filter-map
+;; Filters a map by applying the corresponding filter function to its keys
+;;
+;; Inputs:
+;;   fn: (function) function to be applied to the map keys and/or values
+;;   map: (map) Map to be filtered
+;;
+;; Output: (map) Filtered map
+(define (filter-map fn map)
+    (iterate (out-map {} key (keys map))
+      (if (fn key (get map key))
+          (assoc out-map key (get map key))
+          out-map)))
+
+;; shadow-ds
+;; Extends a dataset adding a shadow field for each field used in input_fields.
+;; If the number of shadow fields is less than 5, fields are repeated.
+;;
+;; Inputs
+;;  ds-id: (dataset-id) Dataset ID
+;;  f-ids: (list) List of field IDs that will be used as input and extended
+;;                with shadow fields
+;;
+;; Output: (list) List of elements [dataset-id shadow-field-ids fields]
+;;   dataset-id: (dataset-id) ID of the dataset with shadow fields
+;;   shadow-fields-ids: (list) List of IDs for the newly added shadow fields
+;;   fields: (map) New fields structure for the fields in the original dataset.
+;;                 This structure is needed to know the IDs in the
+;;                 shadow dataset for the original fields
+;;
+(define (shadow-ds ds-id fields obj-name)
+  (let (f-id-names (reduce (lambda (x y) (assoc x
+                                                y
+                                                (get-in fields [y "name"])))
+                           {}
+                           (keys fields))
```

Comment:
formatting nit: it's usual to put lambda's bodies below the parameter list, just as in defines, and one gets a bit more room: (and, at least to my eyes, it reads a bit better):

```
(let (f-id-names (reduce (lambda (x y)
                           (assoc x y (get-in fields [y "name"])))
                         {}
                         (keys fields))
      obj-f (filter-map (lambda (x _)
                          (= (get-in fields [x "name"]) obj-name))
                        f-id-names)
```